File: 2023/rust/src/bin/d09p1.rs
use std::fs;

fn main() {
    println!("Day 9 - Part 1");

    let str = fs::read_to_string("../_input/day09.txt").unwrap();
    let lines: Vec<&str> = str.lines().collect();

    let total = process(lines);
    println!("Total: {}", total);
}

fn process(lines: Vec<&str>) -> i32 {
    let readings: Vec<Vec<i32>> = lines
        .iter()
        .map(|s| s.split_whitespace().map(str_to_i32).collect())
        .collect();

    let mut total = 0;
    for reading in readings {
        let mut stack = vec![reading];
        let mut i = 0;
        while !stack[i].iter().all(|e| *e == 0) {
            let v = stack[i].windows(2).map(|w| w[1] - w[0]).collect();
            stack.push(v);
            i += 1;
        }

        stack.reverse();
        stack[0].push(0);

        let mut i = 1;
        while i < stack.len() {
            let a = {
                let a = stack.get(i - 1).unwrap();
                *a.last().unwrap()
            };
            let b = {
                let b = stack.get(i).unwrap();
                *b.last().unwrap()
            };

            stack.get_mut(i).unwrap().push(a + b);

            i += 1;
        }

        total += stack.last().unwrap().last().unwrap();
    }

    total
}

fn str_to_i32(str: &str) -> i32 {
    str.parse().unwrap()
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45"
                    .split('\n')
                    .collect()
            ),
            114
        );
    }
}

File: 2023/swift/d09p1.swift
import Foundation

func main() {
    print("Day 9 - Part 1")

    let raw_data = try! NSString(
        contentsOfFile: "../_input/day09.txt", encoding: String.Encoding.utf8.rawValue)
    let data = raw_data as String

    let total = process(String(data))
    print("Total: \(total)")
}

func process(_ string: String) -> Int {
    let readings: [[Int]] = string.split(separator: "\n").map({
        $0.components(separatedBy: " ").map({ Int($0)! })
    })

    var total = 0
    for reading in readings {
        var stack = [reading]

        var i = 0
        while !stack[i].allSatisfy({ $0 == 0 }) {
            let s = stack[i]
            var v: [Int] = []
            var vi = 0

            while vi < s.endIndex - 1 {
                v.append(s[vi + 1] - s[vi])
                vi += 1
            }

            stack.append(v)
            i += 1
        }

        stack.reverse()
        stack[0].append(0)

        i = 1
        while i < stack.count {
            let a = stack[i - 1].last!
            let b = stack[i].last!

            stack[i].append(a + b)
            i += 1
        }

        total += stack.last!.last!
    }

    return total
}

main()
